backend/src/options.mjs

```diff
@@ -0,0 +1,91 @@
+/**
+ * @typedef {'html' | 'csv'} Format
+ */
+
+/**
+ * @typedef {Object} DateRange
+ * @property {Date} start
+ * @property {Date} end
+ */
+
+/**
+ * Supported options when running the program.
+ */
+export class Options {
+  /**
+   * @param {Object} userInput
+   */
+  constructor(userInput) {
+    /**
+     * Output format (HTML or CSV).
+     * @type {Format}
+     */
+    this.format = parseFormat(userInput);
+
+    /**
+     * Date range to use to collect usage of on-demand tables.
+     * @type {DateRange}
+     */
+    this.dateRange = parseDateRange(userInput);
+  }
+
+}
+
+export const OPTION_FORMAT = 'format';
+export const supportedFormats = ['html', 'csv'];
+export const defaultFormat = 'html';
+export const OPTION_FROM_DATE = 'from-date';
+export const OPTION_TO_DATE = 'to-date';
+
+/**
+ * @param {Object} inputs
+ * @return {Format}
+ */
+const parseFormat = (inputs) => {
+  const format = inputs[OPTION_FORMAT];
+  if (format !== undefined) {
+    if (supportedFormats.includes(format)) {
+      return format
+    } else {
+      throw `Unsupported format '${format}'. Valid formats are ${supportedFormats.map(f => `'${f}'`).join(', ')}.`
+    }
+  } else {
+    return defaultFormat
+  }
+};
+
+/**
+ * @param {Object} inputs
+ * @return {DateRange}
+ */
+const parseDateRange = (inputs) => {
+  const fromDate = inputs[OPTION_FROM_DATE];
+  const toDate = inputs[OPTION_TO_DATE];
+  if (fromDate !== undefined && toDate !== undefined) {
+    const start = parseDate(fromDate);
+    const end = parseDate(toDate);
+    if (end <= start) {
+      throw `Invalid combination of parameters: '${OPTION_FROM_DATE}' must be before '${OPTION_TO_DATE}'`;
+    }
+    return { start, end }
+  } else if (fromDate === undefined && toDate === undefined) {
+    const now = new Date();
+    const oneMonthAgo = new Date(now);
+    oneMonthAgo.setUTCDate(oneMonthAgo.getUTCDate() - 30);
+    return { start: oneMonthAgo, end: now }
+  } else {
+    throw `Invalid combination of parameters: '${OPTION_FROM_DATE}' and '${OPTION_TO_DATE}' must be supplied together`;
```

Comment:
Shouldn't throw when only `fromDate` is defined

Reply:
My take is the following: it is the date _range_ that is optional, so users should supply either both the from and to dates, or nothing. Otherwise, I’m afraid this could lead to confusion: if the from date is `x`, do we expect the end date to be `x + 30 days` or “yesterday”? Why would it be legal to supply only `from-date` but illegal to supply only `to-date`?

Reply:
> if the from date is x, do we expect the end date to be `x + 30 days` or “yesterday”? 

"By default, the end date is the date of today." - there's nothing to be confused here.

> Why would it be legal to supply only `from-date` but illegal to supply only `to-date`?

It doesn't have to be illegal, but this open-ended restriction can't scan since the beginning of time. One option would be to default to `to-date - 30d`, but that would break the default semantics, since `now - 30d` != `to-date - 30d`. Hence why it is just simpler to forbid this particular case, plus less prone to oversights.